include/sys/zstd/zstd.h

```diff
@@ -94,6 +97,123 @@ int zfs_zstd_decompress(void *s_start, void *d_start, size_t s_len,
     size_t d_len, int n);
 void zfs_zstd_cache_reap_now(void);
 
+/*
+ * So, the reason we have all these complicated set/get functions is that
+ * originally, in the zstd "header" we wrote out to disk, we used a 32-bit
+ * bitfield to store the "level" (8 bits) and "version" (24 bits).
+ *
+ * Unfortunately, bitfields make few promises about how they're arranged in
+ * memory...
+ *
+ * As a consequence, we wound up with little endian platforms with a layout
+ * like this in memory:
+ *
+ *      32      24      16      8       0
+ *      +-------+-------+-------+-------+
+ *      | level |   0   |    version    |
+ *      +-------+-------+-------+-------+
+ *
+ * ...and then, after being run through BE_32(), serializing this out to
+ * disk:
+ *
+ *      32      24      16      8       0
+ *      +-------+-------+-------+-------+
+ *      |    version    |   0   | level |
+ *      +-------+-------+-------+-------+
+ *
+ * while on big-endian systems, since BE_32() is a noop there, both in
+ * memory and on disk, we wind up with:
+ *
+ *      32      24      16      8       0
+ *      +-------+-------+-------+-------+
+ *      |   0   |    version    | level |
+ *      +-------+-------+-------+-------+
+ *
+ * (The "0" byte is part of the version field, but will remain 0 until
+ * version 6.55.35, so it is useful both for illustration purposes
+ * and for the "get" code below to know which possible layout of the
+ * header we're dealing with.)
+ *
+ * So now we use the BF32_SET macros to get consistent behavior (the LE
+ * encoding, since x86 currently rules the world) across platforms, but
```

Comment:
It looks like the code actually uses the big-endian encoding (e.g. `	hdr->raw_version_level = BE_32(hdr->raw_version_level);`).  Assuming we plan to maintain this "figure out the byteorder" functionality forever, maybe we should continue to write it in host byteorder, so that there's some chance of exercising this searching code even with pools written by the current software?

Reply:
For the sake of precision - I tested this code against the assumption that it writes out bytes to disk in the same order that x86_64 platforms currently do, by attempting to read the bytes written by my modified code from both {x86_64, sparc64} hosts and by unmodified {x86_64,sparc64} hosts on an unpatched copy of git master on x86_64. Both of the former and the obvious one of the latter work.

Would you prefer I refer to the encoding in question as BSWAP_32(LE), to make it clear that it's the swapped representation of LE's view of the world, not the representation that BE platforms natively have, or something else?

I'd prefer to write only one ordering going forward (though I can straightforwardly go add further test cases to the ZTS PR where the above-discussed property of 00 v0 00 LV and vice-versa are true), so that, if we reach the point where this assertion becomes false (e.g. version > 6.55.35), we can reasonably safely stop using this by default a little while before it becomes a problem.